File: uva11504.py
# -*- coding: utf-8 -*-

n = 0
Part = []
Count = 0
Free = []
adj = []
radj = []
Node = []
N = 0

def Init(n):
    global Free, Part, Count, adj, radj, Node
    Free = []
    Part = []
    Count = 0
    adj = []
    radj = []
    Node = []
    for i in range(n):
        Free.append(True)
        Part.append(0)
        adj.append([])
        radj.append([])

def DFS1(u):
    global Free, Node, adj, N
    Free[u] = False
    for i in range(len(adj[u])):
        if adj[u][i] and not Free[i]:
            DFS1(i)
    Node.append(u)

def DFS2(u, p):
    Free[u] = True
    Part[u] = p
    for i in range(len(adj[u])):
        if not Free[i]:
            DFS2(i, p)

def main():
    Case = 0
    m = 0
    u = 0
    v = 0
    Case = int(input())
    Com = []
    var = ""
    result = []
    global Free, Part, Count, adj, radj, Node, N
    while Case:
        var = input()
        divide = var.split(' ')
        N = int(divide[0])
        m = int(divide[1])
        Init(N)
        for i in range(N):
            Com.append(True)
        while m:
            line = input()
            pecas = line.split(' ')
            pos1 = int(pecas[0])
            pos2 = int(pecas[1])
            adj[pos1-1].append(pos2-1)
            radj[pos2-1].append(pos1-1)
            m -= 1

        for i in range(N):
            if Free[i]:
                DFS1(i)
        nPart = 0
        i = len(Node) - 1
        while i<= len(Node) and i > 0:
            if not Free[Node[i]]:
                DFS2(Node[i], nPart)
                nPart += 1
            i -= 1
        print(adj)
        for u in range(N):
            for i in range(N):
                v = adj[u][i]

                print("v %d" % v)
                print("Part[u] %d Part[i] %d" %(Part[u], Part[i]))
                if (Part[u] != Part[i]):
                    Com[Part[v]] = False
        ans = 0
        i = 1
        while i <= nPart:
            if Com[i]:
                ans += 1
            i += 1
        result.append(ans)
        Case -= 1
    for i in result:
        print(i)
    return 0


if __name__ == "__main__":
    main()

File: uva11504.c
#include <stdio.h>


int G[100][100];
int GL[100][100];
int posnum[100], visit[100], Part[100];
int nump=0;
int N = 0;

void Init() {
    visit, true, n+1);
    adj = vector< vector<int> >(n+1);
    radj = vector< vector<int> >(n+1);
    Count = 0;
    Node.clear();
}

void DFS(int v){
    visit[v]=1;
    for(int i=0, i<sizeof(adj[u]); i++){
        if (visit[adj[u][i]]) DFS(adj[u][i]);
        }
    }
    nump++;
    posnum[v] = v;
}

void DFS2(int v, int p) {
    visit[v]=2;
    Part[v] = p;
    int i;
    for (i=0;i<N;i++){
        if (!visit[GL[v][i]]){
            DFS2(GL[v][i], p);
        }
    }
}

int main() {
	int i,j;
    int Case = 0 , n = 0, m = 0, u = 0, v = 0, soma = 0;
    int result[50], pos = 0;
    scanf("%d", &Case);
    while (Case--) {
        scanf("%d %d", &n, &m);
        N = n;

        while (m--) {
            scanf("%d %d", &u, &v);
            G[u][v] = 1;
        }
        for(int i=1; i<=N; i++){
            if(!visit[i]){
                DFS(i);
            }
        }
        for (i=0;i<N;i++){
            for (j=0;j<N;j++){
                GL[i][j]=G[j][i];
            }
        }
        int nPart = 0;
        for (int i=nump; i>=0; i--){
            if (!visit[posnum[i]]) {
                DFS2(posnum[i], ++nPart);
            }
        }
        int Com[nPart+1];
        for(i=0; i<nPart; i++){
            Com[i] = 1;
        }
        for(int u=1; u<=N; u++){
            for (int i=0; i<N; i++) {
                int v = G[u][i];
                if (Part[u]!=Part[v]){
                    Com[Part[v]] = 0;
                }
            }
        }
        int ans = 0;
        for (i=1; i<=nPart; i++){
            if (Com[i]) ans++;
        }
        printf("%d\n", ans);
    }


    return 0;
}
